README.md
# device-coap-c
[![Build Status](https://jenkins.edgexfoundry.org/view/EdgeX%20Foundry%20Project/job/edgexfoundry/job/device-coap-c/job/main/badge/icon)](https://jenkins.edgexfoundry.org/view/EdgeX%20Foundry%20Project/job/edgexfoundry/job/device-coap-c/job/main/) [![GitHub Latest Dev Tag)](https://img.shields.io/github/v/tag/edgexfoundry/device-coap-c?include_prereleases&sort=semver&label=latest-dev)](https://github.com/edgexfoundry/device-coap-c/tags) ![GitHub Latest Stable Tag)](https://img.shields.io/github/v/tag/edgexfoundry/device-coap-c?sort=semver&label=latest-stable) [![GitHub License](https://img.shields.io/github/license/edgexfoundry/device-coap-c)](https://choosealicense.com/licenses/apache-2.0/) [![GitHub Pull Requests](https://img.shields.io/github/issues-pr-raw/edgexfoundry/device-coap-c)](https://github.com/edgexfoundry/device-coap-c/pulls) [![GitHub Contributors](https://img.shields.io/github/contributors/edgexfoundry/device-coap-c)](https://github.com/edgexfoundry/device-coap-c/contributors) [![GitHub Committers](https://img.shields.io/badge/team-committers-green)](https://github.com/orgs/edgexfoundry/teams/device-coap-c-committers/members) [![GitHub Commit Activity](https://img.shields.io/github/commit-activity/m/edgexfoundry/device-coap-c)](https://github.com/edgexfoundry/device-coap-c/commits)

> **Warning**  
> The **main** branch of this repository contains work-in-progress development code for the upcoming release, and is **not guaranteed to be stable or working**.
> It is only compatible with the [main branch of edgex-compose](https://github.com/edgexfoundry/edgex-compose) which uses the Docker images built from the **main** branch of this repo and other repos.
>
> **The source for the latest release can be found at [Releases](https://github.com/edgexfoundry/device-coap-c/releases).**


EdgeX device service for CoAP-based REST protocol

This device service supports both CoAP client and CoAP server. CoAP server allows a 3rd party sensor application to push data into EdgeX via CoAP. CoAP client allows EdgeX to read sensor data through auto-events and send a command to 3rd party sensor application. Like HTTP, CoAP provides REST based access to resources, but CoAP is more compact for use in constrained IoT devices.

The device-coap-c service (_device-coap_ for short) is modeled after the HTTP based [device-rest-go](https://github.com/edgexfoundry/device-rest-go) service, and runs over UDP.

device-coap uses DTLS for secure communication to devices. It is written in C, and relies on the well known [libcoap](https://libcoap.net/) library.

For background on device-coap-c, CoAP and low power wireless, see the [presentation](https://zoom.us/rec/share/N2Uh7C9qScsj32bs0T8aNF4VPPOuFSypnhQp3g2LmSFfOA16giRq9gwqpGvNb1HX.kknLyNV7Rj72mPms?startTime=1602514686000) to the Device Working Group.

## Resources

device-coap creates a parameterized CoAP resource. CoAP server may post data to these resources. CoAP client may initiate auto-events to end device's on these resources. CoAP client may send command to end device's coap-server on these resources. 

```
   /a1r/{deviceName}/{resourceName}
```

- `a1r` is short for "API v1 resource", as defined by device-rest-go.
- `deviceName` refers to a `device` managed by the CoAP device service. For example, `res/devices/devices.json` pre-defines a device named 'd1'.
- `resourceName` refers to a `deviceResource` defined in the device profile, as described in the sub-section below.

Payload data posted to one of these resources is type validated, and the resulting value then is sent into EdgeX via the Device SDK's asynchronous `post_readings` capability.

## Profiles

[example-datatype.json](./res/profiles/example-datatype.json) defines  generic resources for data types. The table below shows the available resource names and correspondence with CoAP attributes. 

For example, the 'int' resource name means that EdgeX provides a CoAP resource, `/a1r/{deviceName}/int`. This resource accepts an integer encoded as text, like `42`.

| resourceName | Type   | CoAP Content-Format|
|---------|--------|---------------------------------------|
| int     | Int32  | text/plain                            |
| float   | Float64| text/plain                            |
| json    | String | application/json                      |

>_Note:_ You must define the Content-Format option in the CoAP POST request. See the _Testing_ section below for example use.


## Configuration

This section describes properties in [configuration.yaml](./res/configuration.yaml) as used by device-coap. See the _Configuration and Registry_ section of the EdgeX documentation for background.

### Driver

Below are the recognized properties for the Driver section, followed by an example. These values are read only when starting the device-coap service. So if you change a property, you must restart the service for the change to take effect.


| Key         | Value                                                                             |
|-------------|-----------------------------------------------------------------------------------|
| CoapBindAddr| Address on which CoAP server listens for devices                                  |
| SecurityMode| DTLS client-server security type. Does not support raw public key or certificates.|


```
Driver:
  # Supports IPv4 or IPv6 if provided by network infrastructure. Use '0.0.0.0'
  # for any IPv4 interface, or '::' for any IPv6 interface.
  CoapBindAddr: 0.0.0.0
  # Choose 'PSK' or 'NoSec'
  SecurityMode: PSK
```

### Secrets

If configured for PSK mode, a key must be stored in the service's secret store:
| Secret name | Value                                                                             |
|-------------|-----------------------------------------------------------------------------------|
| PskKey      | Pre-shared key. Accepts only a single key, ignored in NoSec mode.                 |

For example if using insecure mode (secrets in configuration file):

```
Writable:
  InsecureSecrets:
    CoAP:
      SecretName: psk
      SecretData:
        # Key is up to 16 arbitrary bytes; must be base64 encoded here
        PskKey: ME42aURHZ3Uva0Y0eG9lZw==
```

## Devices
### Devices for CoAP Server 

A pre-defined device 'd1' is supplied. At present no properties for the `other` protocol are defined for a device.

```json
{
  "name": "d1",
  "profileName": "example-datatype",
  "description": "Example generic data type device",
  "labels": [ "coap", "rest" ],
  "protocols": { "other": { } }
}
```

### Devices for CoAP Client

A predefined device 'd2' is supplied. 

```json
{
    "name": "d2",
    "profileName": "example-datatype",
    "description": "Example generic data type device",
    "protocols":
    {
        "COAP": 
        {
            "ED_ADDR": "127.0.0.1",
            "ED_SecurityMode": "PSK",
            "ED_PskKey": "hello123"
        }
    },
    "autoEvents":
    [   
        { "sourceName": "int", "onChange": false, "interval": "30s" }
    ]   
}
```

| Key             | Value                                                        |
| --------------- | ------------------------------------------------------------ |
| ED_ADDR         | Address on which CoAP client initiates request to end device |
| ED_SecurityMode | DTLS client-server security type. Does not support raw public key or certificates. Possible values are PSK/NoSec |
| ED_PskKey       | Pre-shared key. Accepts only a single key, ignored in NoSec mode. |

- Auto-events are supported for the resources mentioned in the profile for example `int` resource. 

## Docker Integration

### Building

You can build a Docker image with the command below from the top level directory of a device-coap checkout.

```
   $ make docker
```

### Compose

Below is an example entry for a docker-compose template with the rest of the EdgeX setup. The CoAP server listens on the default secure port, 5684. It also listens on any interface since the CoAP message likely arrives from an external network. However, it is more secure to use the address for the specific interface for CoAP messaging in your setup.

```
  device-coap:
    image: edgexfoundry/device-coap:0.2-dev
    ports:
      - "127.0.0.1:59988:59988"
      - "0.0.0.0:5684:5684/udp"
    container_name: edgex-device-coap
    hostname: edgex-device-coap
    networks:
      - edgex-network
    environment:
      <<: *common-variables
      Service_Host: edgex-device-coap
    depends_on:
      - metadata
      - data
```

## Testing/Simulation for CoAP Server

You can use simulated data to test this service with libcoap's `coap-client` command line tool. The examples below are organized by the SecurityMode defined in the configuration.

**NoSec**

```
   $ coap-client -m post -t 0 -e 1001 coap://127.0.0.1/a1r/d1/int
```
**PSK**

```
   $ coap-client -m post -u r17 -k 0N6iDGgu/kF4xoeg -t 0 -e 1001 coaps://127.0.0.1/a1r/d1/int
```

  * For DTLS PSK, a CoAP client must include a user identity via the `-u` option as well as the same key the server uses. Presently, the device-coap server does not evaluate the identity, only the key. Also, `coap-client` reads the key as a literal string, so characters must be readable from the command line. Finally, notice the protocol in the address is `coaps`. This protocol uses UDP port 5684 rather than 5683 for protocol `coap`.
  * POSTing a text integer value will set the  `Value` of the `Reading` in EdgeX to the string representation of the value as an `Int32`. The POSTed value is verified to be a valid `Int32` value.
  * A 400 error will be returned if the POSTed value fails the `Int32` type verification.

### Zephyr CoAP client

Also see my Zephyr based [edgex-coap-peer](https://github.com/kb2ma/edgex-coap-peer) repository for a simple CoAP client usable on an IoT device. The client posts integer data for the example profile above, to `/a1r/d1/int`.

### RIOT CoAP client

Also see my RIOT based [riot-edgex-coap-client](https://github.com/kb2ma/riot-edgex-coap-client) repository for a more realistic CoAP client. The client posts a temperature measurement from a sensor every 60 seconds to `/a1r/d1/float`.

## Testing/Simulation for CoAP Client

You can use simulated data to test the CoAP client functionality of this device service using libcoap server. Resources must be handled properly in the coap-server to test CoAP client. The examples below are organized by the ED_SecurityMode defined in devices.json.

**NoSec**

```
$ ./coap-server -A 127.0.0.1
```

**PSK**

```
$ ./coap-server -A 127.0.0.1 -k hello123
```

  * For DTLS PSK, a coap-server must include same key the CoAP client uses. The coap-server reads key as a literal string, so characters must be readable from the command line. 

## Development

This section describes how to build and run a device-coap executable independent from Docker, for development or debugging.

### Building

device-coap depends on libcoap and tinydtls. The [build_deps.sh](scripts/build_deps.sh) script provides a template to build these libraries that you can adapt for use at the command line. `build_deps.sh` is intended for use by the Docker build, so first review [Dockerfile.alpine](scripts/Dockerfile.alpine). Notice that it creates a `/device-coap` directory as a workspace, and then runs `build_deps.sh`. Also keep in mind that a Docker build has full privileges over its container filesystem as it runs.


As with any C based EdgeX device project, device-coap also depends on the EdgeX [C SDK](https://github.com/edgexfoundry/device-sdk-c/blob/master) for its SDK library and headers. Finally, see [build.sh](scripts/build.sh) and [build_debug.sh](scripts/build_debug.sh) to build device-coap itself. These scripts may be invoked via `make build` and `make build-debug` respectively.

### Running

Simply run the generated executable. The example below was built with the `build_debug.sh` script.

```
   $ build/debug/device-coap -cf configuration-native.yaml
```

>_Note:_ `configuration-native.yaml` adapts the contents of `configuration.yaml` for use with a separate device-coap executable.

Run with `-h` to see all command line options.


CHANGELOG.md
## EdgeX CoAP Device Service
[Github repository](https://github.com/edgexfoundry/device-coap-c)

## [v3.1.0] Napa - 2023-11-15 (Only compatible with the 3.x releases)


### 🐛 Bug Fixes

- Update Attribution.txt ([#56](https://github.com/edgexfoundry/device-coap-c/issues/56)) ([52eb205…](https://github.com/edgexfoundry/device-coap-c/commit/52eb20590ed2b0ea5aa136895e2756feac496f8f))


### 🤖 Continuous Integration

- Move to alpine 3.18 base ([398e8cf…](https://github.com/edgexfoundry/device-coap-c/commit/398e8cfcddc771eba213213c442487fdd069ace7))
- Add automated release workflow on tag creation ([ee16241…](https://github.com/edgexfoundry/device-coap-c/commit/ee16241a3a49841d34c25e128ff8448c83ba1518))


<a name="3.0.1"></a>

## v3.0.1 - 2023-08-22

### Features ✨
- Bump to SDK v3.0.2

<a name="3.0.0"></a>

## v3.0.0 - 2023-05-31

### Features ✨
- Bump to SDK v3.0
- Pre-shared key now stored via EdgeX secrets mechanism

<a name="2.3.0"></a>

## v2.3.0 - 2022-11-09

### Features ✨
- Bump to SDK v2.3 ([#43](https://github.com/edgexfoundry/device-coap-c/issues/43))

<a name="2.2.0"></a>

## v2.2.0 - 2022-08-08

### Features ✨
- Bump to SDK v2.2 ([#41](https://github.com/edgexfoundry/device-coap-c/issues/41))
- Upgrade Alpine 3.16 ([#40](https://github.com/edgexfoundry/device-coap-c/issues/40))
- Use secret store for PSK key ([#33](https://github.com/edgexfoundry/device-coap-c/issues/33)) ([#34](https://github.com/edgexfoundry/device-coap-c/issues/34)) ([#402b1dc](https://github.com/edgexfoundry/device-coap-c/commits/402b1dc))

### Continuous Integration 🔄
- Use standard container name in Makefile ([#37](https://github.com/edgexfoundry/device-coap-c/issues/37)) ([#64c820b](https://github.com/edgexfoundry/device-coap-c/commits/64c820b))

<a name="v2.1.0"></a>

## v2.1.0 - - 2021-10-25

### Features ✨
- Bump to SDK v2.1 ([#30](https://github.com/edgexfoundry/device-coap-c/issues/30)) ([#12271f6](https://github.com/edgexfoundry/device-coap-c/commits/12271f6))
- Enhancement of CoAP server device service with CoAP client functionality ([#28](https://github.com/edgexfoundry/device-coap-c/issues/28)) ([#ceb58c0](https://github.com/edgexfoundry/device-coap-c/commits/ceb58c0))

### Bug Fixes 🐛
- default install location in alpine 3.14 ([#26](https://github.com/edgexfoundry/device-coap-c/issues/26)) ([#4a68428](https://github.com/edgexfoundry/device-coap-c/commits/4a68428))
- update all TOML to use quote and not single-quote ([#ddee52f](https://github.com/edgexfoundry/device-coap-c/commits/ddee52f))

### Build 👷
- Upgrade alpine 314 ([#24](https://github.com/edgexfoundry/device-coap-c/issues/24)) ([#644a368](https://github.com/edgexfoundry/device-coap-c/commits/644a368))
- Update alpine base to 3.14 ([#22](https://github.com/edgexfoundry/device-coap-c/issues/22)) ([#ac5b898](https://github.com/edgexfoundry/device-coap-c/commits/ac5b898))

<a name="v2.0.0"></a>

## v2.0.1 - 2021-07-28

### Bug Fixes 🐛

- Pin base Docker image to alpine:3.12 ([#a203b38](https://github.com/edgexfoundry/device-coap-c/commits/a203b38))

## v2.0.0 - 2021-06-22

### Features ✨
- Migrate to newer CSDK supporting EdgeX v2 ([#9](https://github.com/edgexfoundry/device-coap-c/issues/9)) ([#cf36b46](https://github.com/edgexfoundry/device-coap-c/commits/cf36b46))

### Bug Fixes 🐛
- Updates addressing review on PR[#3](https://github.com/edgexfoundry/device-coap-c/issues/3) ([#4](https://github.com/edgexfoundry/device-coap-c/issues/4)) ([#9b0ec4d](https://github.com/edgexfoundry/device-coap-c/commits/9b0ec4d))

### Documentation 📖
- Add badges to readme ([#fd990de](https://github.com/edgexfoundry/device-coap-c/commits/fd990de))

<a name="v0.1.0"></a>
## v0.1.0 - 2021-02-01

[Unreleased]: https://github.com/edgexfoundry/device-coap-c/compare/x.y.z...HEAD
[x.y.z]: https://github.com/edgexfoundry/device-coap-c/compare/v0.1.0...x.y.z


src/c/coap-client.c
/* CoAP cli.t .r thread device service
 *
 * Copyright (c) 2021
 * Sudhamani
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include "coap-client.h"

#include <coap2/coap.h>
#include <errno.h>
#include <float.h>
#include <netdb.h>
#include <signal.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>

#include "coap-util.h"
#include "device-coap.h"
#include "edgex/devices.h"

/* controls input loop */
extern volatile sig_atomic_t quit;
iot_data_t *coap_resp_data = NULL;
static bool coap_resp_rcvd = false;

/*
 * Get End device protocol property, expect 5 arguments:
 * @param[in] protocol structure
 * @param[in] protocol name
 * @param[in] exception ptr
 * @param[in] end device structure variable address to store protocol properties
 * @param[in] coap driver pointer
 * @return success or fail
 */
bool GetEndDeviceProtocolProperties(const devsdk_protocols *protocols,
                                    char *protocol_name, iot_data_t **exception,
                                    end_dev_params *end_dev_params_ptr,
                                    coap_driver *driver) {
  coap_driver *sdk_ctx = (coap_driver *)driver;
  const iot_data_t *props =
      devsdk_protocols_properties(protocols, protocol_name);
  if (props == NULL) {
    *exception = iot_data_alloc_string("No COAP protocol in device address",
                                       IOT_DATA_REF);
    return false;
  }
  const char *params_ptr = iot_data_string_map_get_string(props, "ED_ADDR");
  if (params_ptr == NULL) {
    *exception = iot_data_alloc_string("property in device address missing",
                                       IOT_DATA_REF);
    return false;
  }
  strcpy(end_dev_params_ptr->end_dev_addr, params_ptr);
  iot_log_debug(sdk_ctx->lc, "COAP:End dev addr ptr= %s",
                end_dev_params_ptr->end_dev_addr);

  params_ptr = iot_data_string_map_get_string(props, "ED_SecurityMode");
  if (params_ptr == NULL) {
    *exception = iot_data_alloc_string("property in device address missing",
                                       IOT_DATA_REF);
    return false;
  }
  iot_log_debug(sdk_ctx->lc, "COAP: End dev SecMode ptr= %s", params_ptr);
  end_dev_params_ptr->security_mode = find_security_mode(params_ptr);

  switch (end_dev_params_ptr->security_mode) {
    case SECURITY_MODE_UNKNOWN: {
      iot_log_error(sdk_ctx->lc, "COAP:ED Unknown security mode");
      return false;
    }
    case SECURITY_MODE_PSK: {
      params_ptr = iot_data_string_map_get_string(props, "ED_PskKey");
      if (params_ptr == NULL) {
        *exception = iot_data_alloc_string("property in device address missing",
                                           IOT_DATA_REF);
        return false;
      }
      iot_log_debug(sdk_ctx->lc, "COAP:End dev PskKey ptr= %s", params_ptr);
      if (strlen(params_ptr)) {
        strcpy(end_dev_params_ptr->psk_key, params_ptr);
        iot_log_debug(sdk_ctx->lc, "COAP:PSK key len %u",
                      strlen(end_dev_params_ptr->psk_key));
      } else {
        iot_log_error(sdk_ctx->lc, "COAP:ED PSK key not in configuration");
        return false;
      }
      break;
    }
    case SECURITY_MODE_NOSEC:
      break;
    default:
      break;
  }
  return true;
}
/*
 * coap get request callback handler. Get coap data from the coap pdu and update
 * it to coap_resp_data
 */
static void message_handler(struct coap_context_t *ctx, coap_session_t *session,
                            coap_pdu_t *sent, coap_pdu_t *received,
                            const coap_tid_t id) {
  edgex_device *device = NULL;
  edgex_deviceresource *resource = NULL;
  uint8_t *data = NULL;
  size_t len = 0;
  coap_driver *sdk_ctx = (coap_driver *)impl;
  coap_resp_data = NULL;
  coap_show_pdu(LOG_WARNING, received);
  switch (sent->code) {
    case COAP_REQUEST_GET: {
      /* Validate URI, expect 3 segments: /a1r/{device-name}/{resource-name} */
      if (!parse_path(sent, &device, &resource)) {
        goto finish;
      }

      if (!coap_get_data(received, &len, &data)) {
        iot_log_error(sdk_ctx->lc, "COAP:invalid data of len %u", len);
      }

      iot_log_debug(sdk_ctx->lc, "COAP: coap device resource type %d",
                    resource->properties->type);

      /* Validate and read payload. Content format from option must be
       * acceptable for resource value type. */

      switch (resource->properties->type.type) {
        case IOT_DATA_FLOAT64:
          /* data conversion requires a null terminated string */
          coap_resp_data = read_data_float64(data, len);
          iot_log_debug(sdk_ctx->lc, "COAP:coap float data=%s, len = %d", data,
                        len);
          break;

        case IOT_DATA_INT32:
          iot_log_debug(sdk_ctx->lc, "COAP:coap int32 data=%s, len = %d", data,
                        len);
          coap_resp_data = read_data_int32(data, len);
          break;

        case IOT_DATA_STRING:
          coap_resp_data = read_data_string(data, len);
          iot_log_debug(sdk_ctx->lc, "COAP:coap json data=%s, len = %d", data,
                        len);
          break;

        default:
          iot_log_error(sdk_ctx->lc, "COAP:unsupported resource type %d",
                        resource->properties->type);
          goto finish;
      }
      break;
    }
    case COAP_REQUEST_PUT:
      break;
    default:
      break;
  }

finish:
  if (device != NULL) {
    edgex_free_device(sdk_ctx->service, (edgex_device *)device);
  }
  coap_resp_rcvd = true;
}
/*
Handling NACK messages for coap requests
*/
static void nack_handler(coap_context_t *context, coap_session_t *session,
                         coap_pdu_t *sent, coap_nack_reason_t reason,
                         const coap_tid_t id) {
  coap_driver *sdk_ctx = (coap_driver *)impl;
  switch (reason) {
    case COAP_NACK_TOO_MANY_RETRIES:
    case COAP_NACK_NOT_DELIVERABLE:
    case COAP_NACK_RST:
    case COAP_NACK_TLS_FAILED:
    case COAP_NACK_ICMP_ISSUE:
      coap_resp_rcvd = true;
      iot_log_error(sdk_ctx->lc, "COAP:NACK response from server");
      break;
    default:
      coap_resp_rcvd = true;
      break;
  }
  return;
}

/*
waits for coap response for CON request from end device
*/
static void WaitForCoapResponseFromEndDevice(coap_context_t *ctx,
                                             coap_session_t *session) {
  while (1) {
    coap_io_process(ctx, COAP_IO_WAIT);
    if (coap_resp_rcvd == true || quit == 1) {
      coap_session_release(session);
      coap_free_context(ctx);
      coap_resp_rcvd = false;
      break;
    }
  }
}
/*
send put request to end device
*/
int CoapSendCommandToEndDevice(uint8_t *data, size_t len, char *dev_name,
                               char *resource_name,
                               end_dev_params *end_dev_params_ptr,
                               coap_driver *driver) {
  coap_context_t *ctx = NULL;
  coap_session_t *session = NULL;
  coap_address_t dst;
  coap_pdu_t *pdu = NULL;
  coap_driver *sdk_ctx = (coap_driver *)driver;
  int result = EXIT_FAILURE;
  uint8_t *post_data = data;
  char uri[COAP_URI_MAXLEN + 1] = {0};

  coap_startup();
  coap_proto_t proto = COAP_PROTO_UDP;
  char *port = "5683";

  if (end_dev_params_ptr->security_mode != SECURITY_MODE_NOSEC) {
    proto = COAP_PROTO_DTLS;
    port = "5684";
  }

  if (resolve_address(end_dev_params_ptr->end_dev_addr, port, &dst) < 0) {
    coap_log(LOG_CRIT, "COAP:failed to resolve address\n");
    return result;
  }
  iot_log_debug(sdk_ctx->lc, "COAP: End dev addr = %s",
                end_dev_params_ptr->end_dev_addr);

  iot_log_debug(sdk_ctx->lc, "COAP: Data = %d, Len = %d", *post_data, len);
  /* create CoAP context and a client session */
  ctx = coap_new_context(NULL);
  if (ctx == NULL) {
    iot_log_error(sdk_ctx->lc, "COAP:coap new context creation failed\n");
    return result;
  }
  if (end_dev_params_ptr->security_mode == SECURITY_MODE_PSK) {
    iot_log_debug(sdk_ctx->lc, "COAP-client:ED psk key = %s, len=%d",
                  (uint8_t *)end_dev_params_ptr->psk_key,
                  strlen(end_dev_params_ptr->psk_key));
    if (!ctx || !(session = coap_new_client_session_psk(
                      ctx, NULL, &dst, proto, "r17",
                      (uint8_t *)(end_dev_params_ptr->psk_key),
                      strlen(end_dev_params_ptr->psk_key)))) {
      iot_log_error(sdk_ctx->lc, "COAP:cannot initialize PSK");
      return result;
    }
  } else {
    if (!ctx || !(session = coap_new_client_session(ctx, NULL, &dst, proto))) {
      coap_log(LOG_EMERG, "COAP:cannot create client session\n");
      return result;
    }
  }
  coap_register_response_handler(ctx, message_handler);
  coap_register_nack_handler(ctx, nack_handler);

  /* construct CoAP message */
  pdu = coap_pdu_init(COAP_MESSAGE_CON, COAP_REQUEST_PUT,
                      coap_new_message_id(session) /* message id */,
                      coap_session_max_pdu_size(session));
  if (!pdu) {
    coap_log(LOG_EMERG, "COAP:cannot create PDU\n");
    return result;
  }

  sprintf(uri, "/a1r/%s/%s", dev_name, resource_name);
  int res;
#define BUFSIZE 40
  unsigned char _buf[BUFSIZE] = {0};
  unsigned char *buf = _buf;
  size_t buflen;
  buflen = BUFSIZE;
  res = coap_split_path((const uint8_t *)uri, strlen(uri), buf, &buflen);

  while (res--) {
    coap_add_option(pdu, COAP_OPTION_URI_PATH, coap_opt_length(buf),
                    coap_opt_value(buf));
    iot_log_debug(sdk_ctx->lc, "uri.path_val = %s\n", coap_opt_value(buf));

    buf += coap_opt_size(buf);
  }
  if (result <= 0) {  // size
    result = EXIT_FAILURE;
    return result;
  }
  if (!coap_add_data(pdu, len, (unsigned char *)post_data)) {
    return result;
  }
  coap_show_pdu(LOG_WARNING, pdu);

  /* and send the PDU */
  if (coap_send(session, pdu) == COAP_INVALID_TID) {
    return result;
  }
  WaitForCoapResponseFromEndDevice(ctx, session);
  result = EXIT_SUCCESS;
  return result;
}
/*
send coap get request to end device. waits for the response form end device.
*/
int CoapGetRequestToEndDevice(char *dev_name, char *resource_name,
                              end_dev_params *end_dev_params_ptr,
                              coap_driver *driver) {
  coap_context_t *ctx = NULL;
  coap_session_t *session = NULL;
  coap_address_t dst;
  coap_pdu_t *pdu = NULL;
  coap_driver *sdk_ctx = (coap_driver *)driver;
  int result = EXIT_FAILURE;
  char uri[COAP_URI_MAXLEN + 1] = {0};

  coap_startup();

  coap_proto_t proto = COAP_PROTO_UDP;
  char *port = "5683";

  if (end_dev_params_ptr->security_mode != SECURITY_MODE_NOSEC) {
    proto = COAP_PROTO_DTLS;
    port = "5684";
  }

  if (resolve_address(end_dev_params_ptr->end_dev_addr, port, &dst) < 0) {
    coap_log(LOG_CRIT, "COAP:failed to resolve address\n");
    return result;
  }
  iot_log_debug(sdk_ctx->lc, "COAP: End dev addr = %s",
                end_dev_params_ptr->end_dev_addr);

  /* create CoAP context and a client session */
  ctx = coap_new_context(NULL);
  if (ctx == NULL) {
    iot_log_error(sdk_ctx->lc, "COAP:coap new context creation failed\n");
    return result;
  }
  if (end_dev_params_ptr->security_mode == SECURITY_MODE_PSK) {
    iot_log_debug(sdk_ctx->lc, "COAP-client:ED psk key = %s, len=%d",
                  (uint8_t *)end_dev_params_ptr->psk_key,
                  strlen(end_dev_params_ptr->psk_key));
    if (!ctx || !(session = coap_new_client_session_psk(
                      ctx, NULL, &dst, proto, "r17",
                      (uint8_t *)(end_dev_params_ptr->psk_key),
                      strlen(end_dev_params_ptr->psk_key)))) {
      iot_log_error(sdk_ctx->lc, "COAP:cannot initialize PSK");
      return result;
    }
  } else {
    if (!ctx || !(session = coap_new_client_session(ctx, NULL, &dst, proto))) {
      coap_log(LOG_EMERG, "COAP:cannot create client session\n");
      return result;
    }
  }
  coap_register_response_handler(ctx, message_handler);
  coap_register_nack_handler(ctx, nack_handler);

  /* construct CoAP message */
  pdu = coap_pdu_init(COAP_MESSAGE_CON, COAP_REQUEST_GET,
                      coap_new_message_id(session) /* message id */,
                      coap_session_max_pdu_size(session));
  if (!pdu) {
    coap_log(LOG_EMERG, "COAP:cannot create PDU\n");
    return result;
  }

  sprintf(uri, "/a1r/%s/%s", dev_name, resource_name);
  int res;
#define BUFSIZE 40
  unsigned char _buf[BUFSIZE] = {0};
  unsigned char *buf = _buf;
  size_t buflen;
  buflen = BUFSIZE;
  res = coap_split_path((const uint8_t *)uri, strlen(uri), buf, &buflen);
  while (res--) {
    result = coap_add_option(pdu, COAP_OPTION_URI_PATH, coap_opt_length(buf),
                             coap_opt_value(buf));
    iot_log_debug(sdk_ctx->lc, "uri.path_val = %s\n", coap_opt_value(buf));

    buf += coap_opt_size(buf);
  }
  if (result <= 0) {  // size
    result = EXIT_FAILURE;
    return result;
  }

  coap_show_pdu(LOG_WARNING, pdu);
  /* and send the PDU */
  if (coap_send(session, pdu) == COAP_INVALID_TID) {
    coap_log(LOG_EMERG, "COAP:coap_send cannot send pdu\n");
    return result;
  }
  WaitForCoapResponseFromEndDevice(ctx, session);
  result = EXIT_SUCCESS;
  return result;
}
